### PR TITLE
Update from arkworks 3 --> 4

### PR DESCRIPTION
**Pull request by z-tech**
What does this PR do?

- it updates dependencies in cargo toml from arkworks 3 to 4 and works changes through the code
- see https://github.com/arkworks-rs/ldt/pull/11

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -7,22 +7,21 @@ edition = "2018"
...
+ark-relations = { version = "^0.4.0", default-features = false, optional = true}
 tracing = { version = "0.1", default-features = false, features = [ "attributes" ], optional = true}
 
 [dev-dependencies]
-ark-test-curves = { version = "^0.3.0", default-features = false, features = ["bls12_381_scalar_field", "mnt4_753_scalar_field"] }
+ark-test-curves = { version = "^0.4.2", default-features = false, features = ["bls12_381_scalar_field", "mnt4_753_scalar_field"] }
 
 [patch.crates-io]
-ark-sponge = {git = "https://github.com/arkworks-rs/sponge"}
 ark-r1cs-std = { git = "https://github.com/arkworks-rs/r1cs-std" }
```
Comment (Pratyush) [inline Cargo.toml:23]: We probably want the patch for `ark-crypto-primitives` too.

**Comment by Pratyush** [comment]
Nightly needs to fix this along the lines of this PR: https://github.com/arkworks-rs/algebra/pull/790

**Comment by z-tech** [comment]
> Nightly needs to fix this along the lines of this PR: https://github.com/arkworks-rs/algebra/pull/790

Gotcha. I'm on holiday this week I'll see what I can do Monday.

**Comment by z-tech** [comment]
@Pratyush hi, got nightly working with Giacomo's help.

I was wondering what the purpose of the no_std workflow is? I could to try to work through those errors but there's ~200 of them and I'd like to do this only if there is a specific purpose 😅. LMK if you can. Thanks.

**Comment by z-tech** [comment]
> @Pratyush hi, got nightly working with Giacomo's help.
> 
> I was wondering what the purpose of the no_std workflow is? I could to try to work through those errors but there's ~200 of them and I'd like to do this only if there is a specific purpose 😅. LMK if you can. Thanks.

So for a dependency chain something like `ark_poly -> hashbrown -> ahash -> once_cell -> std` is there some way I could fix this?

**Pull request closed without merging**